envoy/tcp/upstream.h

```diff
@@ -144,6 +144,11 @@ class GenericUpstream {
    * to secure mode. Implemented only by start_tls transport socket.
    */
   virtual bool startUpstreamSecureTransport() PURE;
+
+  /* Called when upstream starttls socket is converted to tls and upstream ssl info
+   *  needs to be set in connection's stream_info.
+   */
+  virtual Ssl::ConnectionInfoConstSharedPtr getUpstreamConnectionSslInfo() PURE;
```

Comment:
I think you can do it a bit simpler and avoid adding another accessor (`getUpstreamConnectionSslInfo`). Just modify `GenericUpstream::startUpstreamSecureTransport` in https://github.com/envoyproxy/envoy/blob/main/envoy/tcp/upstream.h#L146 to return Tcp::ConnectionPool::ConnectionData* (as in https://github.com/envoyproxy/envoy/blob/main/envoy/tcp/upstream.h#L140-L141). Then in https://github.com/envoyproxy/envoy/blob/main/source/common/tcp_proxy/upstream.cc#L46-L50 return valid pointer on success or nullptr on failure.

In `Filter::startUpstreamSecureTransport` you can check the return value and extract ssl() directly.

WDYT?